### ISSUE TITLE
Python: certain 'reimplements' entries missing

### ISSUE THREAD

**Issue by Stewori** [doxygen/doxygen#8480]
**Describe the issue**
Consider in Python a method that overrides a method from a super class. Unlike in e.g. C++ or Java, a method in Python does not need to match the number of arguments of the parent method or be call-compatible at all in order to override/hide the parent method. (One of the things I dislike about Python but it is what it is). This is because the Python interpreter maintains methods in a dictionary for the class, which only maps by name, not considering args.
As an example, take the following program:

```
class A:
    def meth1(self):
        print(1)

    def meth2(self):
        print(2)

class B(A):
    def meth1(self, a):
        print(3)

    def meth2(self):
        print(4)

a = A()
b = B()
a.meth2()
b.meth2()
a.meth1()
b.meth1(0)
b.meth1()
```
The output is
```
2
4
1
3
Traceback (most recent call last):
  File "test.py", line 21, in <module>
    b.meth1()
TypeError: meth1() missing 1 required positional argument: 'a'
```
Note that `b.meth1()` is *not* resolved to the super implementation in `A.meth1` which would work perfectly fine as an arg-less call. This shows that `B.meth1(self, a)` factually reimplements `A.meth1(self)` despite the mismatch in arguments. The supermethod is not longer reachable (without dirty tricks), even if the arguments used in the call target it.
Doxygen does not seem to be aware of this semantics. It recognizes a submethod only as a reimplementation if the argument count matches. The resulting xml snippet has a `reimplements` entry for `meth2` but lacks it for `meth1`:

```
      <memberdef kind="function" id="classtest_1_1B_1a0438c06675f1a365be57159aa7ecc341" prot="public" static="no" const="no" explicit="no" inline="no" virt="non-virtual">
        <type>def</type>
        <definition>def test.B.meth1</definition>
        <argsstring>(self, a)</argsstring>
        <name>meth1</name>
        <param>
          <type>self</type>
          <defname>self</defname>
        </param>
        <param>
          <type><ref refid="namespacetest_1a83aa194b6dd9443622d122efb7d1d8df" kindref="member">a</ref></type>
          <defname>a</defname>
        </param>
        <briefdescription>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file="test.py" line="9" column="1" bodyfile="test.py" bodystart="9" bodyend="11"/>
      </memberdef>
```

**Expected behavior**
I expect the output xml for `meth1` to contain an appropriate `reimplements` entry:

```
      <memberdef kind="function" id="classtest_1_1B_1a0438c06675f1a365be57159aa7ecc341" prot="public" static="no" const="no" explicit="no" inline="no" virt="non-virtual">
        <type>def</type>
        <definition>def test.B.meth1</definition>
        <argsstring>(self, a)</argsstring>
        <name>meth1</name>
        <reimplements refid="classtest_1_1A_1a16dd08060ee2c025b0a633a1279cc1ed">meth1</reimplements>
        <param>
          <type>self</type>
          <defname>self</defname>
        </param>
        <param>
          <type><ref refid="namespacetest_1a83aa194b6dd9443622d122efb7d1d8df" kindref="member">a</ref></type>
          <defname>a</defname>
        </param>
        <briefdescription>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file="test.py" line="9" column="1" bodyfile="test.py" bodystart="9" bodyend="11"/>
      </memberdef>
```

**To Reproduce**
Find the example from above and doxygen config attached. The issue is demonstrated like mentioned above.
After running doxygen on the example, the files in the subfolder ``xml`` contain the snippets shown above.
[issue_python_reimplements.zip](https://github.com/doxygen/doxygen/files/6255549/issue_python_reimplements.zip)

**Version**
Tested with self-compiled clone from 2021-04-04:
```

/blah/doxygen/build/bin/doxygen -v
1.9.2 (97415f7b4504d2be65338464731d7bbb4aa4d993)
```

`lsb_release -a`:
```
LSB Version:    core-9.20160110ubuntu0.2-amd64:core-9.20160110ubuntu0.2-noarch:security-9.20160110ubuntu0.2-amd64:security-9.20160110ubuntu0.2-noarch
Distributor ID:    LinuxMint
Description:    Linux Mint 18 Sarah
Release:    18
Codename:    sarah
```

**Additional context**
I know, from C++ and Java perspective it appears strange that method reimplementation would be determined only by method name and not by the arguments. However, the Python interpreter works that way. It does not know the concept of method overloading and a method hides or overrides any super method of the same name, regardless of its arguments.
For Python code, Doxygen's `reimplements` and likewise the `reimplementedby` entries should work according to that behavior.

**Comment by Stewori** [comment]
Just noticed that the method parameter `a` is linked to the unrelated variable `a` from module top-level.
```
        <param>
          <type><ref refid="namespacetest_1a83aa194b6dd9443622d122efb7d1d8df" kindref="member">a</ref></type>
          <defname>a</defname>
        </param>
```
Interesting. Is this by intention? It appears to be accidental, but it does not really hurt, so I won't bother to file an issue about this. It may be even useful to warn that the parameter masks a global variable.

**Comment by albert-github** [comment]
Regarding the overloading / re-implementing do you have a reference to a pep or similar document describing this behavior?

With the remark about the `a` argument, you mean the reference in xml/classtest_1_1B.xml?
I see that this also happens in html/classtest_1_1B.html
As `a` is here a dummy argument there shouldn't be a link. It indeed will mask a global variable, looks like this also happens in a similar cpp program: 
```
#include <stdio.h>

int ca = 7;

void fie(int &ca)
{
  ca = 9;
}
int main(int argc, char **argv)
{
  int cb = 5;
  fie(cb);
}
```
and is existing for a long time.

In my opinion this is also a bug (and worth an issue).

**Comment by Stewori** [comment]
> Regarding the overloading / re-implementing do you have a reference to a pep or similar document describing this behavior?

It is difficult to find this in the specification. So far I just found the following indicators:

While it is "common knowledge" that Python does not support overloading, the Python manual hardly states this explicitly.
Looking at https://docs.python.org/3/faq/programming.html#how-can-i-overload-constructors-or-methods-in-python it becomes clear that "overloading" is just emulated in Python using optional arguments and default values. I admit this is not really a proof that method lookup is based only on the method name. (If necessary I can show you the name lookup in CPython's source code.)

Then there is this PEP (not accepted but deferred): https://www.python.org/dev/peps/pep-3124
It wanted to add "true" overloading to Python and related functionality. Such a PEP would not have been filed if Python supported overloading in the first place.

"Deferred" means that third party libs can add this functionality. Indeed one can customize method lookup behavior of a class via decorators and custom metaclasses. There are projects that use this to modify method lookup in classes to take args into account. E.g. explained [here](https://www.codementor.io/@arpitbhayani/overload-functions-in-python-13e32ahzqt) and there is e.g. [this](https://pypi.org/project/overloading/) project. Such projects would not exist if Python supported true overloading out of the box.

However, I think Doxygen should assume Python's default behavior without fancy customization.


Finally, one can easily assert the behavior by writing small test programs. In addition to the provided example, one can examine a class' `__dict__` object and will find that the keys are plainly the method names, not containing arguments.

**Comment by albert-github** [comment]
This indeed the problem with python, besides some other weaknesses, at the moment, no standard or authoritative description of how the kernel functions but just a lot of "common knowledge"  and things that are handled by means of extra packages (of which some are available in different incompatible versions).

Most likely doxygen should follow here the rule of "name lookup" (even though this uncommon for other languages).

**Comment by Stewori** [comment]
> no standard or authoritative description of how the kernel functions but just a lot of "common knowledge"

I can tell you a thing or two about it: used to commit to Jython core :-/
At least there are the PEPs, the reference manual and the reference implementation source code.

Give me another try to pin the described behavior in the official reference manual.
In [this section](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy) from the official manual, scroll to the entry "Custom classes". There the following is stated:

> Custom class types are typically created by class definitions (see section Class definitions). A class has a namespace implemented by a dictionary object. Class attribute references are translated to lookups in this dictionary, e.g., `C.x` is translated to `C.__dict__["x"]` (although there are a number of hooks which allow for other means of locating attributes). When the attribute name is not found there, the attribute search continues in the base classes. This search of the base classes uses the C3 method resolution order which behaves correctly even in the presence of ‘diamond’ inheritance structures where there are multiple inheritance paths leading back to a common ancestor.

Note that with "attributes" they also refer to methods: they are talking about the whole class namespace. In Python, methods are really nothing else than certain callable attributes. (This also gets clear because they talk about "method resolution order" in that context.) The next paragraph explicitly states methods as examples for attributes:

> When a class attribute reference (for class C, say) would yield a class method object, it is transformed into an instance method object whose `__self__` attribute is C.

This asserts that they use the term "attributes" like in Doxygen "members" is used. All attribute lookups, includeing method lookups are just done by name:

> `C.x` is translated to `C.__dict__["x"]`



> Most likely doxygen should follow here the rule of "name lookup" (even though this uncommon for other languages).

Exactly.

**Comment by Stewori** [comment]
> With the remark about the a argument, you mean the reference in xml/classtest_1_1B.xml?

Yes. All xml code snippets I posted are from `xml/classtest_1_1B.xml`.

**Comment by Stewori** [comment]
Just noticed that also `listofallmembers` may require an adjustment for this issue. In `classtest_1_1B.xml` it currently is:

```
    <listofallmembers>
      <member refid="classtest_1_1B_1a0438c06675f1a365be57159aa7ecc341" prot="public" virt="non-virtual"><scope>test::B</scope><name>meth1</name></member>
      <member refid="classtest_1_1A_1a16dd08060ee2c025b0a633a1279cc1ed" prot="public" virt="non-virtual" ambiguityscope="test::A."><scope>test::B</scope><name>meth1</name></member>
      <member refid="classtest_1_1B_1a6597b00f6410f4e650cb27d67230e367" prot="public" virt="non-virtual"><scope>test::B</scope><name>meth2</name></member>
    </listofallmembers>
```

When resolving this issue, one might want to erase the entry
```
<member refid="classtest_1_1A_1a16dd08060ee2c025b0a633a1279cc1ed" prot="public" virt="non-virtual" ambiguityscope="test::A."><scope>test::B</scope><name>meth1</name></member>
```
from that list. However, it is easy to recognize and to sort out by looking for `ambiguityscope="test::A."`. So, adjusting that list is not a priority as far as I am concerned. On the other hand, I don't really know the meaning of `ambiguityscope`. Maybe this is unintentional or a manifestation of the name conflict during lookup. Somehow, a part of Doxygen is noticing that something is strange....(?)

**Comment by doxygen** [comment]
I've just committed a possible fix that should better reflect Python's behavior.
For Python code, doxygen will no longer try to match the parameter lists when matching derived members with the same name, and also not try to do the ambiguity resolution that is visible in the list of all members.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
